nexus/src/app/audit_log.rs

```diff
@@ -0,0 +1,124 @@
+        rqctx: &RequestContext<ApiContext>,
+    ) -> CreateResult<AuditLogEntryInit> {
+        let actor = opctx.authn.actor();
+        let entry = AuditLogEntryInit::new(
+            rqctx.request_id.clone(),
+            rqctx.endpoint.operation_id.clone(),
+            rqctx.request.uri().to_string(),
+            rqctx.request.remote_addr().ip(),
+            rqctx
+                .request
+                .headers()
+                .get("User-Agent")
+                .and_then(|value| value.to_str().ok())
+                .map(|s| safe_truncate(s, 255).to_string()),
+            actor.map(|a| a.actor_id()),
+            actor.and_then(|a| a.silo_id()),
+            opctx.authn.scheme_used().map(|s| s.to_string()),
+        );
+        self.db_datastore.audit_log_entry_init(opctx, entry).await
+    }
+
+    /// Complete an existing audit log entry with result info like end time,
+    /// HTTP status code, error message, etc. Note we retry write failures
+    /// because we really want this to go through.
+    pub(crate) async fn audit_log_entry_complete<R: HttpResponse>(
+        &self,
+        opctx: &OpContext,
+        entry: &AuditLogEntryInit,
+        result: &Result<R, HttpError>,
+    ) -> UpdateResult<()> {
+        let (status_code, error_code, error_message) = match result {
+            Ok(response) => (response.status_code(), None, None),
+            Err(error) => (
+                error.status_code.as_status(),
+                error.error_code.clone(),
+                // For now we only log the external message. If we want to log
+                // the internal message, we need to make sure that is ok to do.
+                Some(error.external_message.clone()),
+            ),
+        };
+
+        let update = AuditLogCompletion::new(
+            status_code.as_u16(),
+            error_code,
+            error_message,
+        );
+
+        // Should retry at roughly 250ms, 750ms, 1750ms (plus however long the
+        // tries take). We really want this write to go through.
+        let backoff = backoff::ExponentialBackoffBuilder::new()
+            .with_multiplier(2.0)
+            .with_initial_interval(Duration::from_millis(250))
+            .with_max_elapsed_time(Some(Duration::from_secs(5)))
+            .build();
+
+        let mut count = 0;
+        backoff::retry_notify(
+            backoff,
+            || async {
+                self.db_datastore
+                    .audit_log_entry_complete(opctx, &entry, update.clone())
+                    .await
+                    .map_err(backoff::BackoffError::transient)
+            },
+            |error, delay| {
+                let id = entry.id;
+                count += 1;
+                error!(
+                    self.log,
+                    "failed to complete audit log entry {id}. retry {count} in {delay:?}";
+                    "error" => ?error
+                );
+            },
+        )
+        .await
```

Comment:
Not sure how to test this in an integration test, but it did work as expected when I hard coded `return Err(...)` into the datastore function and ran the tests.

<img width="1100" height="202" alt="Image" src="https://github.com/user-attachments/assets/a40d21f1-9dda-4692-ba57-75337878ba74" />